### PR TITLE
Add an unstable --json=unused-externs flag to print unused externs

### PR DESCRIPTION
**Pull request by est31**
This adds an unstable flag to print a list of the extern names not used by cargo.

This PR will enable cargo to collect unused dependencies from all units and provide warnings.
The companion PR to cargo is: https://github.com/rust-lang/cargo/pull/8437

The goal is eventual stabilization of this flag in rustc as well as in cargo.

Discussion of this feature is mostly contained inside these threads: #57274 #72342 #72603

The feature builds upon the internal datastructures added by #72342

Externs are uniquely identified by name and the information is sufficient for cargo.
If the mode is enabled, rustc will print json messages like:

```
{"unused_extern_names":["byteorder","openssl","webpki"]}
```

For a crate that got passed byteorder, openssl and webpki dependencies but needed none of them.

### Q: Why not pass -Wunused-crate-dependencies?
A: See [ehuss's comment here](https://github.com/rust-lang/rust/issues/57274#issuecomment-624839355)
   TLDR: it's cleaner. Rust's warning system wasn't built to be filtered or edited by cargo.
   Even a basic implementation of the feature would have to change the "n warnings emitted" line that rustc prints at the end.
   Cargo ideally wants to synthesize its own warnings anyways. For example, it would be hard for rustc to emit warnings like
   "dependency foo is only used by dev targets", suggesting to make it a dev-dependency instead.

### Q: Make rustc emit used or unused externs?
A: Emitting used externs has the advantage that it simplifies cargo's collection job.
   However, emitting unused externs creates less data to be communicated between rustc and cargo.
   Often you want to paste a cargo command obtained from `cargo build -vv` for doing something
   completely unrelated. The message is emitted always, even if no warning or error is emitted.
   At that point, even this tiny difference in "noise" matters. That's why I went with emitting unused externs.

### Q: One json msg per extern or a collective json msg?
A: Same as above, the data format should be concise. Having 30 lines for the 30 crates a crate uses would be disturbing to readers.
   Also it helps the cargo implementation to know that there aren't more unused deps coming.

### Q: Why use names of externs instead of e.g. paths?
A: Names are both sufficient as well as neccessary to uniquely identify a passed `--extern` arg.
   Names are sufficient because you *must* pass a name when passing an `--extern` arg.
   Passing a path is optional on the other hand so rustc might also figure out a crate's location from the file system.
   You can also put multiple paths for the same extern name, via e.g. `--extern hello=/usr/lib/hello.rmeta --extern hello=/usr/local/lib/hello.rmeta`,
   but rustc will only ever use one of those paths.
   Also, paths don't identify a dependency uniquely as it is possible to have multiple different extern names point to the same path.
   So paths are ill-suited for identification.

### Q: What about 2015 edition crates?
A: They are fully supported.
   Even on the 2015 edition, an explicit `--extern` flag is is required to enable `extern crate foo;` to work (outside of sysroot crates, which this flag doesn't warn about anyways).
   So the lint would still fire on 2015 edition crates if you haven't included a dependency specified in Cargo.toml using `extern crate foo;` or similar.
   The lint won't fire if your sole use in the crate is through a `extern crate foo;`   statement, but that's not its job.
   For detecting unused `extern crate foo` statements, there is the `unused_extern_crates` lint
   which can be enabled by `#![warn(unused_extern_crates)]` or similar.

cc @jsgf @ehuss @petrochenkov @estebank

### CODE REVIEW

File: src/librustc_session/config.rs [hunk 1]
```diff
@@ -1104,11 +1105,12 @@ pub fn parse_color(matches: &getopts::Matches) -> ColorConfig {
 ///
 /// The first value returned is how to render JSON diagnostics, and the second
 /// is whether or not artifact notifications are enabled.
-pub fn parse_json(matches: &getopts::Matches) -> (HumanReadableErrorType, bool) {
+pub fn parse_json(matches: &getopts::Matches) -> (HumanReadableErrorType, bool, bool) {
```
Comment (jsgf) [inline src/librustc_session/config.rs:1108]: Should `(HumanReadableErrorType, bool, bool)` become a real named type? Two `bool` is pretty ungainly.

File: src/librustc_session/options.rs [hunk 2]
```diff
@@ -142,6 +142,9 @@ top_level_options!(
         // by the compiler.
         json_artifact_notifications: bool [TRACKED],
 
+        // `true` if we're emitting a JSON blob containing the unused externs
+        json_unused_externs: bool [TRACKED],
```
Comment (jsgf) [inline src/librustc_session/options.rs:146]: Does this need to be tracked?

Comment (est31) [reply src/librustc_session/options.rs:146]: I wanted to be on the safe side but i guess it can be UNTRACKED.

File: src/librustc_session/config.rs (hunk 1 above)
Comment (est31) [reply src/librustc_session/config.rs:1108]: Good point.

**Comment by jsgf** [comment]
This is an interesting approach. I can see a possible path to making Buck support this, but it would need some philosophical discussion about how concerns should be separated (ie, is it really up to Buck to diagnose unused dependencies?).

**Review by davidtwco — APPROVED** [review]
Implementation looks good to me.

**Comment by ehuss** [comment]
I haven't had a chance to test and review this in depth, but one minor thing I think would be good to discuss is adding some kind of tag to the JSON messages so they can be easily distinguished. For example, Cargo uses the `"reason"` field to distinguish them, and I was wondering if rustc should adopt something similar.

> but it would need some philosophical discussion about how concerns should be separated (ie, is it really up to Buck to diagnose unused dependencies?).

Can you say more? I was assuming #72342 would be sufficient for your use case, and Buck wouldn't need to use this extra flag.

**Comment by est31** [comment]
Yeah currently cargo tries to deserialize each message by its type one after another https://github.com/rust-lang/cargo/blob/0506d8d5c9dd229c94641dd39a37bd947dc0a770/src/cargo/core/compiler/mod.rs#L1192

One could put everything into an enum that's internally tagged. This would create a tag value for each of the messages. Tools that invoke rustc can then do with it whatever they want. Ignore it, or create an enum of their own. https://serde.rs/enum-representations.html#internally-tagged

If there's support, I can make a PR to add an enum to rustc.

**Comment by jsgf** [comment]
@ehuss Well, it would be ideal if we can make one mechanism work well for everything. As I see it the pros and cons of the two approaches are:

-Wunused-crate-dependencies:
+ fits in with the general Rust lint handing flow, including deny/allow/warn/forbid
+ straightforward to integrate with external workflows (just another warning)
+ easy to integrate with a build system, so long as dependencies are precise
- doesn't work well with Cargo's model
- requires --extern-location to be able to provide good diagnostics/tooling

--json=unused-externs:
+ leaves decisions to build system, so can apply higher-level logic
+ allows build system to provide diagnostics
- not treated as a rustc lint (Cargo doesn't currently have a lint concept)
- requires close coupling between rustc and build system

Strictly speaking, --json=unused-externs isn't necessary if you have -Wunused-crate-dependencies, since in principle Cargo could parse the diagnostics output and look for the appropriate messages and consume/process/filter them. But it does leave awkwardness like the number of warnings being wrong unless Cargo also corrects those which is messy.

**Comment by est31** [comment]
Huh, tests still failing.

**Comment by camelid** [comment]
@est31 Can you fix the failing tests or do you want a new review?

**Comment by est31** [comment]
Yeah I need to find time to debug and fix the failing tests.

File: compiler/rustc_metadata/src/creader.rs [hunk 3]
```diff
@@ -184,6 +186,27 @@ impl CStore {
     crate fn has_global_allocator(&self) -> bool {
         self.has_global_allocator
     }
+
+    pub fn report_unused_deps(&self, tcx: TyCtxt<'_>) {
+        // We put the check for the option before the lint_level_at_node call
+        // because the call mutates internal state and introducing it
+        // leads to some ui tests failing.
```
Comment (Mark-Simulacrum) [inline compiler/rustc_metadata/src/creader.rs:193]: Can you say more about these UI test failures? Maybe we should bless them instead?

File: compiler/rustc_session/src/config.rs [hunk 4]
```diff
@@ -1718,6 +1738,10 @@ pub fn build_session_options(matches: &getopts::Matches) -> Options {
     let mut debugging_opts = build_debugging_options(matches, error_format);
     check_debug_option_stability(&debugging_opts, error_format, json_rendered);
 
+    if !debugging_opts.unstable_options && json_unused_externs {
+        early_error(error_format, "--json=unused-externs is unstable: -Z unstable-options");
```
Comment (Mark-Simulacrum) [inline compiler/rustc_session/src/config.rs:1742]: Can we use the standard message here? https://github.com/rust-lang/rust/blob/8e8939b8045b7af6076fb718e2e298844aaf4650/compiler/rustc_session/src/config.rs#L1980-L1987

Fine to just duplicate it, though.

File: src/librustdoc/doctest.rs [hunk 5]
```diff
@@ -162,6 +170,44 @@ pub fn run(options: Options) -> Result<(), ErrorReported> {
         Some(testing::Options::new().display_output(display_warnings)),
     );
 
+    // Collect and warn about unused externs, but only if we've gotten
+    // reports for each doctest
+    if json_unused_externs {
+        let unused_extern_reports: Vec<_> =
+            std::mem::take(&mut unused_extern_reports.lock().unwrap());
+        if unused_extern_reports.len() == compiling_test_count {
```
Comment (Mark-Simulacrum) [inline src/librustdoc/doctest.rs:178]: Why do we have this condition here? Is it for when some of the tests e.g. failed to parse? I think a comment would be good.

File: src/librustdoc/doctest.rs [hunk 6]
```diff
@@ -231,6 +277,14 @@ impl DirState {
     }
 }
 
+#[derive(serde::Serialize, serde::Deserialize)]
+struct UnusedExterns {
```
Comment (Mark-Simulacrum) [inline src/librustdoc/doctest.rs:281]: Can we at least leave a comment here that this should match the rustc definition? (And equivalently on the rustc struct?) Ideally we would deduplicate these...

File: src/librustdoc/doctest.rs [hunk 7]
```diff
@@ -272,6 +327,12 @@ fn run_test(
     if as_test_harness {
         compiler.arg("--test");
     }
+    if options.json_unused_externs && !compile_fail {
```
Comment (Mark-Simulacrum) [inline src/librustdoc/doctest.rs:330]: Why do we care if it is going to fail to compile? Do we not support printing unused externs in that case? (It seems reasonable to still, e.g.

```rust
extern crate hyper;

fn bar() { 
 let x: () = 5; // fails to compile, but we don't use hyper
}
```

File: compiler/rustc_metadata/src/creader.rs (hunk 3 above)
Comment (est31) [reply compiler/rustc_metadata/src/creader.rs:193]: I can't remember exactly which UI test failed but it was unrelated and I didn't want the feature to change stuff unless it's really enabled.

Comment (Mark-Simulacrum) [reply compiler/rustc_metadata/src/creader.rs:193]: I feel uncomfortable landing this without knowing more :)

File: src/librustdoc/doctest.rs (hunk 6 above)
Comment (est31) [reply src/librustdoc/doctest.rs:281]: I've wondered about deduplicating the definitions but they differ by ownership, and didn't felt the benefits of deduplication justified the additional copies.

File: src/librustdoc/doctest.rs (hunk 7 above)
Comment (est31) [reply src/librustdoc/doctest.rs:330]: If something fails to compile, it might not fully run the stage of the compiler that would use the external crate. So hyper might *appear* to be used to folks but the compiler won't recognize it because of the compilation error. It might only see some crates, it might see no crates at all. All bets are off. So that's why I made compile fail cases not count into crate usages. If your external crate is only used in compile fail cases of rustdoc snippets, then maybe removing it is a good idea :).

Regarding your example, note that `extern crate hyper` marks the crate used by this lint just as any `pub use` or `use` reexport would :), see PR description.

File: src/librustdoc/doctest.rs (hunk 6 above)
Comment (est31) [reply src/librustdoc/doctest.rs:281]: Also note that it also appears inside cargo. A "real" solution would involve building a common crate for the json format that cargo, rustc, and rustdoc can import.

File: compiler/rustc_metadata/src/creader.rs (hunk 3 above)
Comment (est31) [reply compiler/rustc_metadata/src/creader.rs:193]: I'll rebase and remove the last commit, then it'll show the failure.

File: src/librustdoc/doctest.rs (hunk 5 above)
Comment (est31) [reply src/librustdoc/doctest.rs:178]: Yes, in some instances rustc doesn't print the unused externs report. The check is here to catch this. IIRC there is a similar one on the cargo implementation side.

File: compiler/rustc_metadata/src/creader.rs (hunk 3 above)
Comment (est31) [reply compiler/rustc_metadata/src/creader.rs]: @Mark-Simulacrum see the error logs. The entire list is very long. This is an excerpt:

<details>

```

The actual stderr differed from the expected stderr.
Actual stderr saved to /checkout/obj/build/x86_64-unknown-linux-gnu/test/ui/lint/reasons-forbidden/reasons-forbidden.stderr
To update references, rerun the tests and pass the `--bless` flag
To only update this specific test, also pass `--test-args lint/reasons-forbidden.rs`

error: 1 errors occurred comparing output.
status: exit code: 1
command: "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2/bin/rustc" "/checkout/src/test/ui/lint/reasons-forbidden.rs" "-Zthreads=1" "--target=x86_64-unknown-linux-gnu" "--error-format" "json" "-Zui-testing" "-Zdeduplicate-diagnostics=no" "-Zemit-future-incompat-report" "--emit" "metadata" "-C" "prefer-dynamic" "--out-dir" "/checkout/obj/build/x86_64-unknown-linux-gnu/test/ui/lint/reasons-forbidden" "-A" "unused" "-Crpath" "-O" "-Cdebuginfo=0" "-Zunstable-options" "-Lnative=/checkout/obj/build/x86_64-unknown-linux-gnu/native/rust-test-helpers" "-L" "/checkout/obj/build/x86_64-unknown-linux-gnu/test/ui/lint/reasons-forbidden/auxiliary"
stdout:
------------------------------------------

------------------------------------------
stderr:
------------------------------------------
error[E0453]: allow(unsafe_code) overruled by outer forbid(unsafe_code)
  --> /checkout/src/test/ui/lint/reasons-forbidden.rs:19:13
   |
LL |     unsafe_code,
   |     ----------- `forbid` level set here
...
LL |     #[allow(unsafe_code)]
   |             ^^^^^^^^^^^ overruled by previous forbid
   |
   = note: our errors & omissions insurance policy doesn't cover unsafe Rust

error[E0453]: allow(unsafe_code) overruled by outer forbid(unsafe_code)
  --> /checkout/src/test/ui/lint/reasons-forbidden.rs:19:13
   |
LL |     unsafe_code,
   |     ----------- `forbid` level set here
...
LL |     #[allow(unsafe_code)]
   |             ^^^^^^^^^^^ overruled by previous forbid
   |
   = note: our errors & omissions insurance policy doesn't cover unsafe Rust

error[E0453]: allow(unsafe_code) overruled by outer forbid(unsafe_code)
  --> /checkout/src/test/ui/lint/reasons-forbidden.rs:19:13
   |
LL |     unsafe_code,
   |     ----------- `forbid` level set here
...
LL |     #[allow(unsafe_code)]
   |             ^^^^^^^^^^^ overruled by previous forbid
   |
   = note: our errors & omissions insurance policy doesn't cover unsafe Rust

error[E0453]: allow(unsafe_code) overruled by outer forbid(unsafe_code)
  --> /checkout/src/test/ui/lint/reasons-forbidden.rs:19:13
   |
LL |     unsafe_code,
   |     ----------- `forbid` level set here
...
LL |     #[allow(unsafe_code)]
   |             ^^^^^^^^^^^ overruled by previous forbid
   |
   = note: our errors & omissions insurance policy doesn't cover unsafe Rust

error[E0453]: allow(unsafe_code) overruled by outer forbid(unsafe_code)
  --> /checkout/src/test/ui/lint/reasons-forbidden.rs:19:13
   |
LL |     unsafe_code,
   |     ----------- `forbid` level set here
...
LL |     #[allow(unsafe_code)]
   |             ^^^^^^^^^^^ overruled by previous forbid
   |
   = note: our errors & omissions insurance policy doesn't cover unsafe Rust

error[E0453]: allow(unsafe_code) overruled by outer forbid(unsafe_code)
  --> /checkout/src/test/ui/lint/reasons-forbidden.rs:19:13
   |
LL |     unsafe_code,
   |     ----------- `forbid` level set here
...
LL |     #[allow(unsafe_code)]
   |             ^^^^^^^^^^^ overruled by previous forbid
   |
   = note: our errors & omissions insurance policy doesn't cover unsafe Rust

error[E0453]: allow(unsafe_code) overruled by outer forbid(unsafe_code)
  --> /checkout/src/test/ui/lint/reasons-forbidden.rs:19:13
   |
LL |     unsafe_code,
   |     ----------- `forbid` level set here
...
LL |     #[allow(unsafe_code)]
   |             ^^^^^^^^^^^ overruled by previous forbid
   |
   = note: our errors & omissions insurance policy doesn't cover unsafe Rust

error: aborting due to 7 previous errors

For more information about this error, try `rustc --explain E0453`.

------------------------------------------


---- [ui] ui/tool_lints.rs stdout ----
diff of stderr:

16	LL | #[warn(foo::bar)]
17	   |        ^^^
18	
-	error: aborting due to 3 previous errors
+	error[E0710]: an unknown tool name found in scoped lint: `foo::bar`
+	  --> $DIR/tool_lints.rs:1:8
+	   |
+	LL | #[warn(foo::bar)]
+	   |        ^^^
+	
+	error: aborting due to 4 previous errors
20	
21	For more information about this error, try `rustc --explain E0710`.
22	


The actual stderr differed from the expected stderr.
Actual stderr saved to /checkout/obj/build/x86_64-unknown-linux-gnu/test/ui/tool_lints/tool_lints.stderr
To update references, rerun the tests and pass the `--bless` flag
To only update this specific test, also pass `--test-args tool_lints.rs`

error: 1 errors occurred comparing output.
status: exit code: 1
command: "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2/bin/rustc" "/checkout/src/test/ui/tool_lints.rs" "-Zthreads=1" "--target=x86_64-unknown-linux-gnu" "--error-format" "json" "-Zui-testing" "-Zdeduplicate-diagnostics=no" "-Zemit-future-incompat-report" "--emit" "metadata" "-C" "prefer-dynamic" "--out-dir" "/checkout/obj/build/x86_64-unknown-linux-gnu/test/ui/tool_lints" "-A" "unused" "-Crpath" "-O" "-Cdebuginfo=0" "-Zunstable-options" "-Lnative=/checkout/obj/build/x86_64-unknown-linux-gnu/native/rust-test-helpers" "-L" "/checkout/obj/build/x86_64-unknown-linux-gnu/test/ui/tool_lints/auxiliary"
stdout:
------------------------------------------

------------------------------------------
stderr:
------------------------------------------
error[E0710]: an unknown tool name found in scoped lint: `foo::bar`
  --> /checkout/src/test/ui/tool_lints.rs:1:8
   |
LL | #[warn(foo::bar)]
   |        ^^^

error[E0710]: an unknown tool name found in scoped lint: `foo::bar`
  --> /checkout/src/test/ui/tool_lints.rs:1:8
   |
LL | #[warn(foo::bar)]
   |        ^^^

error[E0710]: an unknown tool name found in scoped lint: `foo::bar`
  --> /checkout/src/test/ui/tool_lints.rs:1:8
   |
LL | #[warn(foo::bar)]
   |        ^^^

error[E0710]: an unknown tool name found in scoped lint: `foo::bar`
  --> /checkout/src/test/ui/tool_lints.rs:1:8
   |
LL | #[warn(foo::bar)]
   |        ^^^

error: aborting due to 4 previous errors

For more information about this error, try `rustc --explain E0710`.

------------------------------------------


---- [ui] ui/unknown-lint-tool-name.rs stdout ----
diff of stderr:

34	LL | #[allow(foo::bar)]
35	   |         ^^^
36	
-	error: aborting due to 6 previous errors
+	error[E0710]: an unknown tool name found in scoped lint: `foo::bar`
+	  --> $DIR/unknown-lint-tool-name.rs:1:9
+	   |
+	LL | #![deny(foo::bar)]
+	   |         ^^^
+	
+	error[E0710]: an unknown tool name found in scoped lint: `foo::bar`
+	  --> $DIR/unknown-lint-tool-name.rs:5:9
+	   |
+	LL | #[allow(foo::bar)]
+	   |         ^^^
+	
+	error: aborting due to 8 previous errors
38	
39	For more information about this error, try `rustc --explain E0710`.
40	


The actual stderr differed from the expected stderr.
Actual stderr saved to /checkout/obj/build/x86_64-unknown-linux-gnu/test/ui/unknown-lint-tool-name/unknown-lint-tool-name.stderr
To update references, rerun the tests and pass the `--bless` flag
To only update this specific test, also pass `--test-args unknown-lint-tool-name.rs`

error: 1 errors occurred comparing output.
status: exit code: 1
command: "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2/bin/rustc" "/checkout/src/test/ui/unknown-lint-tool-name.rs" "-Zthreads=1" "--target=x86_64-unknown-linux-gnu" "--error-format" "json" "-Zui-testing" "-Zdeduplicate-diagnostics=no" "-Zemit-future-incompat-report" "--emit" "metadata" "-C" "prefer-dynamic" "--out-dir" "/checkout/obj/build/x86_64-unknown-linux-gnu/test/ui/unknown-lint-tool-name" "-A" "unused" "-Crpath" "-O" "-Cdebuginfo=0" "-Zunstable-options" "-Lnative=/checkout/obj/build/x86_64-unknown-linux-gnu/native/rust-test-helpers" "-L" "/checkout/obj/build/x86_64-unknown-linux-gnu/test/ui/unknown-lint-tool-name/auxiliary"
stdout:
------------------------------------------

------------------------------------------
stderr:
------------------------------------------
error[E0710]: an unknown tool name found in scoped lint: `foo::bar`
  --> /checkout/src/test/ui/unknown-lint-tool-name.rs:1:9
   |
LL | #![deny(foo::bar)] //~ ERROR an unknown tool name found in scoped lint: `foo::bar`
   |         ^^^

error[E0710]: an unknown tool name found in scoped lint: `foo::bar`
  --> /checkout/src/test/ui/unknown-lint-tool-name.rs:5:9
   |
LL | #[allow(foo::bar)] //~ ERROR an unknown tool name found in scoped lint: `foo::bar`
   |         ^^^

error[E0710]: an unknown tool name found in scoped lint: `foo::bar`
  --> /checkout/src/test/ui/unknown-lint-tool-name.rs:1:9
   |
LL | #![deny(foo::bar)] //~ ERROR an unknown tool name found in scoped lint: `foo::bar`
   |         ^^^

error[E0710]: an unknown tool name found in scoped lint: `foo::bar`
  --> /checkout/src/test/ui/unknown-lint-tool-name.rs:5:9
   |
LL | #[allow(foo::bar)] //~ ERROR an unknown tool name found in scoped lint: `foo::bar`
   |         ^^^

error[E0710]: an unknown tool name found in scoped lint: `foo::bar`
  --> /checkout/src/test/ui/unknown-lint-tool-name.rs:1:9
   |
LL | #![deny(foo::bar)] //~ ERROR an unknown tool name found in scoped lint: `foo::bar`
   |         ^^^

error[E0710]: an unknown tool name found in scoped lint: `foo::bar`
  --> /checkout/src/test/ui/unknown-lint-tool-name.rs:5:9
   |
LL | #[allow(foo::bar)] //~ ERROR an unknown tool name found in scoped lint: `foo::bar`
   |         ^^^

error[E0710]: an unknown tool name found in scoped lint: `foo::bar`
  --> /checkout/src/test/ui/unknown-lint-tool-name.rs:1:9
   |
LL | #![deny(foo::bar)] //~ ERROR an unknown tool name found in scoped lint: `foo::bar`
   |         ^^^

error[E0710]: an unknown tool name found in scoped lint: `foo::bar`
  --> /checkout/src/test/ui/unknown-lint-tool-name.rs:5:9
   |
LL | #[allow(foo::bar)] //~ ERROR an unknown tool name found in scoped lint: `foo::bar`
   |         ^^^

error: aborting due to 8 previous errors

For more information about this error, try `rustc --explain E0710`.

------------------------------------------
```

</details>

Summary of failures:

```
 failures:
    [ui] ui/deduplicate-diagnostics.rs#duplicate
    [ui] ui/error-codes/E0452.rs
    [ui] ui/error-codes/E0453.rs
    [ui] ui/error-codes/E0602.rs
    [ui] ui/feature-gates/feature-gate-lint-reasons.rs
    [ui] ui/feature-gates/feature-gate-unsafe_block_in_unsafe_fn.rs
    [ui] ui/lint/crate_level_only_lint.rs
    [ui] ui/lint/lint-forbid-attr.rs
    [ui] ui/lint/lint-forbid-cmdline.rs
    [ui] ui/lint/lint-malformed.rs
    [ui] ui/lint/lint-unknown-lint-cmdline.rs
    [ui] ui/lint/outer-forbid.rs
    [ui] ui/lint/reasons-erroneous.rs
    [ui] ui/lint/reasons-forbidden.rs
    [ui] ui/tool_lints.rs
    [ui] ui/unknown-lint-tool-name.rs
```

Comment (est31) [reply compiler/rustc_metadata/src/creader.rs]: Maybe if #78848 is merged it'll be ok.

Comment (est31) [reply compiler/rustc_metadata/src/creader.rs]: Nope, not fixed by  #78848. Just rebased on top of it and it still fails, just now in the llvm-9 job instead of the llvm-8 one.

Comment (Mark-Simulacrum) [reply compiler/rustc_metadata/src/creader.rs]: Ah, ok, I see the cause  -- some things are deduplicated in normal compilation but not here (we force deduplication off for UI testing). Seems fine then.

**Comment by crlf0710** [comment]
@est31 Ping from triage, would you mind resolving the merge conflicts? Thanks.

**Comment by est31** [comment]
@crlf0710 done!

File: compiler/rustc_session/src/config.rs (hunk 4 above)
Comment (Mark-Simulacrum) [reply compiler/rustc_session/src/config.rs:1739]: I think this still isn't implemented.

File: src/librustdoc/doctest.rs (hunk 6 above)
Comment (Mark-Simulacrum) [reply src/librustdoc/doctest.rs:280]: I think we're still missing a comment here just noting that this is duplicated elsewhere. I agree on the real solution but I don't think we're there yet and it's a separate PR at least :)

**Comment by Mark-Simulacrum** [comment]
I left some comments on a few things that didn't get resolved from a previous review it looks like, but they're replies or so. If you can't find them just re-set the review label with triagebot and I can leave them anew or so.

**Comment by JohnCSimon** [comment]
@est31 Ping from triage - unfortunately there's still merge conflicts, can you please resolve them and ping the reviewer? Thanks.

**Comment by JohnCSimon** [comment]
@est31 Ping again from triage - unfortunately there's still merge conflicts, can you please resolve them and switch the tag to S-waiting-on-review? Thanks.

**Comment by JohnCSimon** [comment]
@est31 Ping again from triage - unfortunately there's still merge conflicts, can you please resolve them and switch the tag to S-waiting-on-review? Thanks.

**Comment by est31** [comment]
I've rebased and addressed all review comments. re-r? @Mark-Simulacrum 

@rustbot label: +S-waiting-on-review -S-waiting-on-author

**Review by estebank — APPROVED** [review]

**Comment by Mark-Simulacrum** [comment]
@bors r+ 

Sorry for the delay here. This looks good to me.

**Comment by jsgf** [comment]
BTW I've been using this in some experiments and it works out really well in our new build system.
